### PR TITLE
Use docker-compose build + reorganize Dockerfile layers

### PR DESCRIPTION
**Pull request by cdrini**
## Description
Refactor. Simplify docker build by:
- Use `docker-compose build` commands where possible
- Reorganize Dockerfile layers so pip/npm installation is only required when `requirements*.txt` or `package*.json` changes are made. This means rebuilding the images will be much faster if only a random file in the repo has been changed.
- Updated docs

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -73,9 +73,11 @@ If you are using Docker for Mac or Docker for Windows (or the older Docker Toolb
...
-- Adding or changing core dependencies => you will most likely need to rebuild both `olbase` and `oldev` images. This shouldn't happen too frequently. If you are making this sort of change, you will know exactly what you are doing ;)
+- **Editing python files or web templates?** Simply save the file; gunicorn will auto-reload it.
+- **Editing frontend css or js?** Run `docker-compose exec web npm run-script build-assets`. This will re-generate the assets in the persistent `ol-build` volume mount (so the latest changes will be available between stopping / starting  `web` containers). Note, if you want to view the generated output you will need to attach to the container (`docker-compose exec web bash`) to examine the files in the volume, not in your local dir.
```
Comment (hornc) [inline docker/README.md]: Is this correct? `docker-compose exec web npm run-script build-assets`  I thought the Make tasks handled the building of assets (they always have in the past, and I wasn't aware of a change in the process). This looks like it is adding another path. Do they do the same thing? If they do, we should only have one way to trigger them.

File: docker/README.md [hunk 2]
```diff
@@ -37,8 +37,8 @@ docker-machine start default # Start the docker daemon
 
 # build images
 docker build -t olbase:latest -f docker/Dockerfile.olbase . # 30+ min (Win10Home/Dec 2018)
-docker build -t oldev:latest -f docker/Dockerfile.oldev .   # 10+ min (Win10Home/Dec 2018)
-docker build -t olsolr:latest -f docker/Dockerfile.olsolr . # 5+ min (Win10Home/Dec 2018)
+docker-compose build web # 10+ min (Win10Home/Dec 2018)
```
Comment (hornc) [inline docker/README.md]: I'm not a fan of using docker-compose as a build tool. It doesn't give us we we really want of being able to build multiple images in sequence (which is why olbase is build differently). I thought we had touched on this briefly in a community call a while ago. A shell build script or a Make task would be better and combine all the actions in a simpler to type command. Alternatively, pick and use a current docker build tool to do the job. I believe docker-compose's build functionality is deprecated too (but I can't find a reference ATM).

File: docker/README.md (hunk 1 above)
Comment (cdrini) [reply docker/README.md]: I thought I saw this in the docs somewhere, but can't find it anymore. @jdlrobson what's the recommended command, `make js css` or `npm run-script build-assets`?

File: docker/README.md (hunk 2 above)
Comment (cdrini) [reply docker/README.md]: Haha, I thought I just recently learned about the build option in the compose config, but it's entirely possible that I forgot. I think the docker-compose command is easier to remember? As someone who has to use these commands a lot, I got tired of having to open our docker docs every time I needed to rebuild. I think a make task would be a good idea, but it adds another layer of duplication (do we have just one command? What if I just want to build `web` (which is the only one that usually needs rebuilding)?). I think we should practice (my new favourite word:) [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) with regards to sequential builds; let's cross that bridge when we get to it :P

**Pull request merged**